drivers: rtc: add MAX31331 RTC Support

frmanabat opened the pull request:
Add MAX31331 RTC driver

The MAX31331 is an ultra-low-power real-time clock (RTC) that provides timekeeping with extremely low current consumption (65 nA). It supports automatic switchover to a backup power supply during power-fail events and includes a trickle charger for maintaining external supercapacitor or rechargeable battery backup.

This driver supports the MAX31331’s key features, including:

1. Alarm functionality
2. Timestamping for event detection

More Info: https://www.analog.com/media/en/technical-documentation/data-sheets/MAX31331.pdf

bjarki-andreasen reviewed (commented):
Please extend the [RTC build_all test suite](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all) with this RTC and if you have not already, run it through the [rtc API test suite](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/rtc/rtc_api) in both the local and global thread configurations

frmanabat committed:
drivers: rtc: add MAX31331 RTC Support

Add MAX31331 RTC driver, an ultra-low-power real-time clock (RTC)
that provides timekeeping with
extremely low current consumption (65 nA).

Signed-off-by: Francis Roi Manabat <francisroi.manabat@analog.com>

frmanabat committed:
tests: rtc: include adi,max31331 in build_all test suite

Add the MAX31331 RTC device to the build_all RTC test suite. This ensures
it compiles on all supported platforms without requiring actual hardware.

Signed-off-by: Francis Roi Manabat <francisroi.manabat@analog.com>

frmanabat commented:
v5

1. Fix native_sim build warnings

bjarki-andreasen commented:
> > Please extend the [RTC build_all test suite](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/build_all) with this RTC and if you have not already, run it through the [rtc API test suite](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/rtc/rtc_api) in both the local and global thread configurations
> 
> Hi, I would like to ask how should I proceed with testing in rtc_api? Thanks!

Build and run the RTC test suite for your board which has the rtc connected to it, add an overlay which points to the rtc using an alias:
```
/ {
	aliases {
		rtc = &rtc; /* or whatever its called on your board */
	};
};
```
```
west build -p -b <your board> zephyr/tests/drivers/rtc/rtc_api -DCONFIG_RTC_ALARM=y
```
the test results will be printed to the console when you run the test.

bjarki-andreasen reviewed (dismissed):
Nice work! The changes to the test suite make sense, you are welcome to commit those upstream as well

frmanabat commented:
v6

1. Changed bytes number MACRO to ARRAY_SIZE
2. Adjust set and get time to conform with rtc_time struct

MaureenHelm reviewed (dismissed):
> I did test this via a sample app for the Nucleo U575ZI-Q and the [MAX31331 Shield](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/max31331shld.html#eb-overview).

Can you send a follow-up PR to add the shield?
https://docs.zephyrproject.org/latest/hardware/porting/shields.html

kartben reviewed (changes requested):
please address copilot comments - thanks!
